utilities/replace_versions.py

```diff
@@ -0,0 +1,108 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import os
+import re
+
+version_update_marker = re.compile(r'\{x-version-update:([^:]+):([^}]+)\}')
+version_update_start_marker = re.compile(r'\{x-version-update-start:([^:]+):([^}]+)\}')
+version_update_end_marker = re.compile(r'\{x-version-update-end\}')
+version_regex_str = r'\d+\.\d+\.\d+(?:-\w+)?(?:-\w+)?'
+version_regex = re.compile(version_regex_str)
+
+class CodeModule:
+    def __init__(self, module_str):
+        (name, released, current) = module_str.split(':')
+        self.name = name
+        self.released = released
+        self.current = current.strip()
+
+def replace_versions(version_map, target):
+    newlines = []
+    repl_open = False
+    repl_thisline = False
+    with open(target) as f:
+        for line in f:
+            repl_thisline = repl_open
+            match = version_update_marker.search(line)
+            if match:
+                module_name = match.group(1)
+                version_type = match.group(2)
+                repl_thisline = True
+            else:
+                match = version_update_start_marker.search(line)
+                if match:
+                    module_name = match.group(1)
+                    version_type = match.group(2)
+                    repl_open = True
+                    repl_thisline = True
+                else:
+                    match = version_update_end_marker.search(line)
+                    if match:
+                        repl_open = False
+                        repl_thisline = False
+
+            if repl_thisline:
+                if module_name not in version_map:
+                    raise ValueError('module not found in version.txt: ' + module_name)
+                module = version_map[module_name]
+                new_version = ''
+                if version_type == 'current':
+                    new_version = module.current
+                elif version_type == 'released':
+                    new_version = module.released
+                else:
+                    raise ValueError('invalid version type: ' + version_type)
+
+                newline = re.sub(version_regex_str, new_version, line)
+                newlines.append(newline)
+            else:
+                newlines.append(line)
+
+            if not repl_open:
+                module_name = ''
+                version_type = ''
+
+    with open(target, 'w') as f:
+        for line in newlines:
+            f.write(line)
+
+def replace_versions_all(target):
+    version_map = {}
+    with open('versions.txt') as f:
+        for line in f:
+            if line.startswith('#'):
+                continue
+            module = CodeModule(line)
+            version_map[module.name] = module
+
+    if target:
+        replace_versions(version_map, target)
+    else:
+        for root, dirs, files in os.walk("."):
```

Comment:
Please mention in the doc that the file must be executed from repo root like: 
```
> python utilities/replace_versions.py
```
otherwise this would not work

Reply:
Will do.